telemetry/metrics/metrics.go

```diff
@@ -0,0 +1,149 @@
+	return nil
+}
+
+// UnaryClientMetricsInterceptor returns an unary client grpc interceptor which adds recorder to the grpc request context
+func UnaryClientMetricsInterceptor(recorder MetricsRecorder) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		recorderCtx := NewContextWithRecorder(ctx, recorder)
+		return invoker(recorderCtx, method, req, reply, cc, opts...)
+	}
+}
+
+// StreamClientMetricsInterceptor returns a stream client grpc interceptor which adds recorder to the grpc stream context
+func StreamClientMetricsInterceptor(recorder MetricsRecorder) grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		recorderCtx := NewContextWithRecorder(ctx, recorder)
+		stream, err := streamer(recorderCtx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		return &observedClientStream{
+			ClientStream:    stream,
+			metricsRecorder: recorder,
+		}, nil
+	}
+}
+
+// observedClientStream is a grpc client stream with an attached MetricsRecorder
+type observedClientStream struct {
+	grpc.ClientStream
+	metricsRecorder MetricsRecorder
+}
+
+// *observedClientStream.Context returns the client stream context with the MetricsRecorder attached
+func (o *observedClientStream) Context() context.Context {
+	ctx := o.ClientStream.Context()
+	ctx = NewContextWithRecorder(ctx, o.metricsRecorder)
+	return ctx
+}
+
+// UnaryServerMetricsInterceptor returns an unary server grpc interceptor which adds recorder to the grpc request context
+func UnaryServerMetricsInterceptor(recorder MetricsRecorder) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		recorderCtx := NewContextWithRecorder(ctx, recorder)
+		return handler(recorderCtx, req)
+	}
+}
+
+// StreamServerMetricsInterceptor returns a stream server grpc interceptor which adds recorder to the grpc stream context
+func StreamServerMetricsInterceptor(recorder MetricsRecorder) grpc.StreamServerInterceptor {
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		stream := &observedStream{
+			ServerStream:    ss,
+			metricsrecorder: recorder,
+		}
+		err := handler(srv, stream)
+		return err
+	}
+}
+
+// observedStream is a grpc server stream with an attached MetricsRecorder
+type observedStream struct {
+	grpc.ServerStream
+	metricsrecorder MetricsRecorder
+}
+
+// *observedStream.Context returns the server stream context attached with a MetricsRecorder
+func (o *observedStream) Context() context.Context {
+	// Get the stream context and make sure our recorder is attached.
+	ctx := o.ServerStream.Context()
```

Comment:
Hmm, I thought that the context used for the stream was derived from our parent context in main(). I guess it isn't? How does logging work if the context isn't getting propogated?

Reply:
Context used for the grpc stream isn't derived from main. 

Looking at https://github.com/Snowflake-Labs/sansshell/blob/a5c11d0807beb83407eb2172da38948b6e33ae35/cmd/proxy-server/server/server.go#L462 grpc.Serve() doesn't accept context as argument, so context from main gets obsolete as soon as the server starts serving gRPC. 
I briefly looked at the grpc server API. I don't think there's a way to pass context into a grpc server.

We're doing similar thing for logger - interceptors do the job of inserting logger to the context. 
https://github.com/Snowflake-Labs/sansshell/blob/main/telemetry/telemetry.go#L172

Reply:
Ah, that makes the context solution less clean than I'd hoped. Keeping logging and metrics similar still seems better overall, but just barely.